README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
simtimer
========

simtimer in a little R package designed to simplify (and speeding up) calculating time intervals in discrete event simulations.

Discrete event simulations is a simulation paradigm that is based on the evaluation of events taking place in a time-specific order. Therefore a discrete event simulation calculates many time intervals. simtimer handles dates and times as integers. This makes working with time intervals as easy (and as fast) as subtracting integers.

Installation
------------

You can install simtimer from `CRAN` or `github` with:

``` r
install.packages("simtimer")

# install.packages("devtools")
devtools::install_github("ims-fhs/simtimer")
```

Example
-------

Simtimer allows transformation between datetimes (POSIXt) and sim\_datetimes (integer) with `sim_datetime()` and `datetime()`.

``` r
origin_date <- as.POSIXct("2016-01-01 00:00:00", tz = "UTC")
my_datetime <-  as.POSIXct("2016-01-02 01:01:01", tz = "UTC")
my_simdatetime <- as.sim_datetime(my_datetime, origin_date)
my_simdatetime
#> [1] 90061
```

``` r
as.datetime(my_simdatetime, origin_date)
#> [1] "2016-01-02 01:01:01 UTC"
```

Simtimer allows to manipulate sim\_datetimes and extract parts of sim\_datetimes with `sim_time()`, `sim_wday()` and `sim_date()`.

``` r
sim_time(my_simdatetime)
#> [1] 3661
sim_wday(my_simdatetime, origin_date)
#> [1] "6"
sim_date(my_simdatetime)
#> [1] 1
```


NEWS.md
# simtimer 4.0.0

* Adapt return_value of `simtimer::sim_wday()` to a character, giving the weekday number 
("1" = Monday, "2" = Tuesday, ..., "7" = Sunday). Before this function returned the 
abbreviated weekday (depending on `base::Sys.getlocale()`). The new solution therefore improves platform independency.
* decoupling of lubridate package by replacement with base functions. 


# simtimer 3.0.0

* Added a `NEWS.md` file to track changes to the package.
* Basic functionality as described in readme.md and vignette.



tests/testthat/helper-data.R
# Helper file. Set global test parameters available in all testthat sections
# Use helper-synthetic_data for synthetic data like data.frames, ...


origin_date <- lubridate::ymd_hms("2016-01-01 00:00:00")
t_20160101 <- lubridate::ymd_hms("2016-01-01 00:00:00") # = origin_date!
t_20160102 <- lubridate::ymd_hms("2016-01-02 00:00:00")
t_20170101 <- lubridate::ymd_hms("2017-01-01 00:00:00")

# Set upper bounds for microbenchmark tests. Set units via
# microbenchmark::microbenchmark(function(), times = nL, unit = "us"):
t_500us <- 500 # 500 us



tests/testthat/test-simtimer.R
# context("uuid-check for testdata")
# test_that("uuid of testdata didn't change", {
#   # skip("Skip for development of tests")
#
#   expect_equal(data$uuid, "ba6b7622-452d-11e6-8bc9-c94b5972ecfb")
#   expect_equal(class(data$uuid), "character")
# })

context("basic functions")
test_that("sim_datetime works", {
  # skip("Skip for development of tests")
  expect_equal(as.sim_datetime(t_20160101, origin_date), 0)
  expect_equal(as.sim_datetime(t_20160102, origin_date), 86400)
  expect_equal(as.sim_datetime(t_20170101, origin_date), 366*24*60*60)
})

test_that("datetime works", {
  # skip("Skip for development of tests")

  expect_equal(as.datetime(as.sim_datetime(t_20160101, origin_date), origin_date), t_20160101)
  expect_equal(as.datetime(as.sim_datetime(t_20160102, origin_date), origin_date), t_20160102)
  expect_equal(as.datetime(as.sim_datetime(t_20170101, origin_date), origin_date), t_20170101)
})

test_that("sim_date works", {
  # skip("Skip for development of tests")

  expect_equal(sim_date(0), 0)
  expect_equal(sim_date(24*3600), 1)
  expect_equal(sim_date(365*24*3600), 365)
})

test_that("sim_time delivers the right number of seconds of the day", {
  # skip("Skip for development of tests")

  expect_equal(sim_time(0), 0)
  expect_equal(sim_time(24*60*60), 0)
  expect_equal(sim_time(365*24*60*60), 0)
  expect_equal(sim_time(365*24*60*60 - 1), 24*60*60 - 1)
})

test_that("sim_wday works", {
  # skip("Skip for development of tests")

  expect_equal(sim_wday(0, origin_date), weekdays(origin_date, abbreviate = TRUE))
  expect_equal(sim_wday(24*60*60 - 1, origin_date), weekdays(origin_date + 24*60*60 - 1, abbreviate = TRUE))
  expect_equal(sim_wday(24*60*60, origin_date), weekdays(origin_date + 24*60*60, abbreviate = TRUE))
  expect_equal(sim_wday(6*24*60*60, origin_date), weekdays(origin_date + 6*24*60*60, abbreviate = TRUE))
  expect_equal(sim_wday(365*24*60*60, origin_date), weekdays(origin_date + 365*24*60*60, abbreviate = TRUE))
})



R/simtimer.R
# simtimer converts datetimes to a relative timescale based on seconds for the
# usage inside Discrete-Event Simulation (DES) engines. DES engines have to calculate
# many timestamps and intervalls during a simulation run. The representation of
# datetimes as integers helps to speed up these calculations.
#
# * Functions called sim_something() refer to the relative integer timescale called "sim_datetime".
# * The functions as.sim_datetime() and as.datetime() transform datetimes to sim_datetimes and back.


#' Date part of a sim_datetime
#'
#' sim_date() returns the date part of a sim_datetime.
#' Therefore sim_date() calculates the number of days (24h-intervals) that have
#' passed since origin_date. If the origin_date of sim_datetime has a time component
#' different than 00:00:00, the 24h-intervals are correlated to this particular time component.
#'
#' @param sim_datetime A sim_datetime (integer representing the passed seconds since origin_date)
#'
#' @return the number of days (24h-intervals) that have passed since origin_date
#' @export
#' @examples
#' sim_date(24*60*60-1)
#' # [1] 0
#' sim_date(24*60*60)
#' # [1] 1
#' sim_date(452*24*60*60)
#' # [1] 452
#' origin_date <- as.POSIXct("2016-01-01 00:00:00", tz = "UTC")
#' sim_date(as.sim_datetime(as.POSIXct("2016-01-02 00:01:00", tz = "UTC"), origin_date))
#' # [1] 1
sim_date <- function(sim_datetime) {
  return(trunc(sim_datetime/(24*60*60)))
}

#' Time part of a sim_datetime
#'
#' sim_time() returns the time of a sim_datetime in seconds.
#' The beginning of a day is defined by the time component of origin_date which defines the
#' parameter sim_datetime.
#'
#' @param sim_datetime A sim_datetime (integer representing the passed seconds since origin_date)
#'
#' @return time in seconds (Range: 0-(24*60*60-1))
#' @export
#' @examples
#' sim_time(200)
#' # [1] 200
#' sim_time(24*60*60-1)
#' # [1] 86399
#' sim_time(24*60*60)
#' # [1] 0
#' origin_date <- as.POSIXct("2016-01-01 00:00:00", tz = "UTC")
#' sim_time(as.sim_datetime(as.POSIXct("2016-01-01 00:01:00", tz = "UTC"), origin_date))
#' # [1] 60
#' sim_time(as.sim_datetime(as.POSIXct("2016-01-02 00:01:00", tz = "UTC"), origin_date))
#' # [1] 60
sim_time <- function(sim_datetime) {
  return(sim_datetime %% (24*60*60))
}


#' Weekday part of a sim_datetime
#'
#' sim_wday() returns the weekday of a sim_datetime. It's crucial to use the same origin_date for
#' sim_wday() than the origin_date that was used to generate the sim_datetime.
#' sim_wday() uses the base R format(x, "\%u") function.
#'
#' @param sim_datetime A sim_datetime (integer representing the passed seconds since origin_date)
#' @param origin_date A datetime (POSIXt)
#'
#' @return A character, giving the weekday number ("1" = Monday, "2" = Tuesday, ..., "7" = Sunday)
#' @export
#' @examples
#' origin_date <- as.POSIXct("2016-01-01 00:00:00", tz = "UTC")
#' sim_wday(60, origin_date)
#' sim_wday(3600,origin_date)
#' sim_wday(36*3600,origin_date)
sim_wday <- function(sim_datetime, origin_date) {
  return(as.character(format(origin_date + sim_datetime, "%u")))
}

#' Transformation from a datetime to a sim_datetime
#'
#' as.sim_datetime() transforms a regular datetime element (POSIXt) to a sim_datetime
#' (integer representing the passed seconds since origin_date).
#' The timezone (tz) will be ignored at the moment.
#' Therefore tz of datetime and origin_date should be identical.
#'
#' @param datetime A datetime (POSIXt)
#' @param origin_date A datetime (POSIXt)
#'
#' @return A sim_datetime
#' @export
#'
#' @examples
#' origin_date <- as.POSIXct("2016-01-01 00:00:00", tz = "UTC")
#' as.sim_datetime(as.POSIXct("2016-01-01 00:01:00", tz = "UTC"), origin_date)
#' # [1] 60
#' as.sim_datetime(as.POSIXct("2016-01-02 00:01:00", tz = "UTC"), origin_date)
#' # [1] 86460
as.sim_datetime <- function(datetime, origin_date) {
  return(as.integer(as.numeric(datetime - origin_date, units = "secs")))
}

#' Back-transformation from a sim_datetime to a datetime
#'
#' as.datetime() transforms a sim_datetime element (integer) to a regular datetime
#' element (POSIXt)
#'
#' @param sim_datetime A sim_datetime (integer representing the passed seconds since origin_date)
#' @param origin_date A datetime (POSIXt)
#'
#' @return datetime A POSIXt
#' @export
#' @examples
#' origin_date <- as.POSIXct("2016-01-01 00:00:00", tz = "UTC")
#' as.datetime(60, origin_date)
#' # [1] "2016-01-01 00:01:00 UTC"
#' as.datetime(600, origin_date)
#' # [1] "2016-01-01 00:10:00 UTC"
#' as.datetime(as.sim_datetime(as.POSIXct("2016-01-02 00:00:00", tz = "UTC"), origin_date),
#' origin_date)
#' # [1] "2016-01-02 UTC"
as.datetime <- function(sim_datetime, origin_date) {
  return(origin_date + sim_datetime)
}

